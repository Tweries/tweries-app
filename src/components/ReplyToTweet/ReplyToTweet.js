import classnames from 'classnames';
import React, { useEffect, useState } from 'react';
import fetchTweet from '../../api/fetchTweet';

const copy = {
  'Optional: reply to Tweet URL': 'Optional: reply to Tweet URL'
};

function ReplyToTweet({ onChangeId, onChangeUrl, userId, value }) {
  const [id, setId] = useState(null);
  const [waiting, setWaiting] = useState(false);

  useEffect(() => {
    async function fetchData() {
      try {
        const { statusId } = await fetchTweet({
          replyToTweetUrl: value,
          userId
        });
        if (statusId) {
          setWaiting(false);
          setId(statusId);
        } else {
          setWaiting(false);
          setId(null);
        }
      } catch (error) {
        setWaiting(false);
        setId(null);
      }
    }
    if (userId && value) {
      setWaiting(true);
      fetchData();
    }
  }, [userId, value]);

  useEffect(() => {
    onChangeId(id);
  }, [id, onChangeId]);

  return (
    <textarea
      className={classnames('border border-gray-500 p-2 rounded', {
        'bg-gray-200': waiting === true || value === '',
        'bg-green-200': waiting === false && value !== '' && id !== null,
        'bg-red-200': waiting === false && value !== '' && id === null
      })}
      data-testid="reply-to"
      onChange={({ target: { value } }) => {
        // TODO: add a debounce
        onChangeUrl(value);
      }}
      placeholder={copy['Optional: reply to Tweet URL']}
      rows={1}
      value={value}
    />
  );
}

export default ReplyToTweet;
