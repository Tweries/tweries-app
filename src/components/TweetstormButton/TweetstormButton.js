import { faSpinner } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

const copy = {
  Tweet: 'Tweet'
};

function TweetstormButton({ disabled, onClick, waiting }) {
  return (
    <button
      className={classnames(
        'bg-gray-300 border border-gray-500 font-bold my-4 px-6 py-2 rounded self-center',
        {
          'tweries-background-color-blue-button': !(disabled || waiting),
          'cursor-auto': disabled || waiting
        }
      )}
      data-testid="tweet"
      disabled={disabled || waiting}
      onClick={onClick}
      type="button"
    >
      {waiting ? <FontAwesomeIcon icon={faSpinner} spin /> : copy.Tweet}
    </button>
  );
}

TweetstormButton.propTypes = {
  disabled: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
  waiting: PropTypes.bool.isRequired
};

export default TweetstormButton;
