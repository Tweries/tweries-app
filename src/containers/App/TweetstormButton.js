import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';
import classnames from 'classnames';
import React from 'react';

const copy = {
  Tweet: 'Tweet'
};

// TODO: move to components
function TweetstormButton({ disabled, onClick, waiting }) {
  return (
    <button
      disabled={disabled}
      className={classnames(
        'bg-gray-300 border border-gray-500 font-bold px-4 mb-2 rounded self-center',
        {
          'bg-blue-600 text-white': !disabled,
          'cursor-auto': disabled
        }
      )}
      data-testid="tweet"
      onClick={onClick}
    >
      {waiting ? <FontAwesomeIcon icon={faSpinner} spin={true} /> : copy.Tweet}
    </button>
  );
}

export default TweetstormButton;
