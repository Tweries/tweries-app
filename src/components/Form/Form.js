import PropTypes from 'prop-types';
import React from 'react';
import { MAX_LENGTH } from '../../constants';
import Counter from '../Counter/Counter';
import ReplyToTweet from '../ReplyToTweet/ReplyToTweet';
import Textarea from '../Textarea/Textarea';
import TweetstormButton from '../TweetstormButton/TweetstormButton';

const copy = {
  'Edits can be made in the boxes below before publishing':
    'Edits can be made in the boxes below before publishing',
  'Start typing. To insert a break prior to reaching 280 characters, please use Newline(s)':
    'Start typing. To insert a break prior to reaching 280 characters, please use Newline(s)',
  Tweet: 'Tweet',
  'Type your thoughts here': 'Type your thoughts here'
};

function Form({
  healthy,
  inReplyToTweetUrl,
  items,
  onChangeSource,
  onChangeTweet,
  onClick,
  replyToTweetCallback,
  setInReplyToTweetUrl,
  source,
  userId,
  waiting
}) {
  function disabled() {
    return (
      !healthy ||
      !items.length > 0 ||
      items.some((item) => item.tweet.length > MAX_LENGTH)
    );
  }

  return (
    <form
      className="flex flex-col"
      onSubmit={/* istanbul ignore next */ (e) => e.preventDefault()}
    >
      <ReplyToTweet
        callback={replyToTweetCallback}
        onChange={setInReplyToTweetUrl}
        tweetUrl={inReplyToTweetUrl}
        userId={userId}
      />
      <p className="italic py-4 text-sm">
        {
          copy[
            'Start typing. To insert a break prior to reaching 280 characters, please use Newline(s)'
          ]
        }
      </p>
      <textarea
        className="p-2 tweries-background-color-blue-white tweries-border"
        data-testid="source"
        name="source"
        onChange={onChangeSource}
        placeholder={copy['Type your thoughts here']}
        rows={8}
        value={source}
      />
      <Counter length={source.length} />
      {items.length > 0 && [
        <p className="italic py-4 text-sm" key="copy">
          {copy['Edits can be made in the boxes below before publishing']}
        </p>,
        <ul className="flex flex-col" data-testid="list" key="list">
          {items.map((item, index) => (
            <li className="flex flex-col" key={item.id}>
              <label className="pb-1 text-sm" htmlFor={item.id}>
                {`${copy.Tweet} #${index + 1}`}
              </label>
              <Textarea item={item} onChange={(e) => onChangeTweet(e, item)} />
              <Counter length={item.tweet.length} type="tweet" />
            </li>
          ))}
        </ul>
      ]}
      <TweetstormButton
        disabled={disabled()}
        onClick={onClick}
        waiting={waiting}
      />
    </form>
  );
}

Form.propTypes = {
  healthy: PropTypes.bool.isRequired,
  inReplyToTweetUrl: PropTypes.string.isRequired,
  items: PropTypes.array.isRequired,
  onChangeSource: PropTypes.func.isRequired,
  onChangeTweet: PropTypes.func.isRequired,
  onClick: PropTypes.func.isRequired,
  replyToTweetCallback: PropTypes.func.isRequired,
  setInReplyToTweetUrl: PropTypes.func.isRequired,
  source: PropTypes.string.isRequired,
  userId: PropTypes.string.isRequired,
  waiting: PropTypes.bool.isRequired
};

export default Form;
