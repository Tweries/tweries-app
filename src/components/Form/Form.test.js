import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import ReplyToTweet from '../ReplyToTweet/ReplyToTweet';
import TweetstormButton from '../TweetstormButton/TweetstormButton';
import Form from './Form';

jest.mock('../../components/ReplyToTweet/ReplyToTweet');
jest.mock('../../components/TweetstormButton/TweetstormButton');

describe('<Form />', () => {
  beforeEach(() => {
    ReplyToTweet.mockImplementation(() => <div>ReplyToTweet</div>);
    TweetstormButton.mockImplementation(() => <div>TweetstormButton</div>);
  });

  it('to match snapshot', () => {
    const { container } = render(<Form healthy items={[]} source="" />);
    expect(container).toMatchSnapshot();
  });

  it('onChangeSource', () => {
    const onChangeSource = jest.fn();

    const { getByTestId } = render(
      <Form healthy items={[]} onChangeSource={onChangeSource} source="" />
    );

    fireEvent.change(getByTestId('source'), {
      target: { value: 'Surely crypto must be able to save us, no?' }
    });

    expect(onChangeSource).toBeCalled();
  });

  it('onChangeTweet', () => {
    const onChangeTweet = jest.fn();

    const { getByTestId } = render(
      <Form
        healthy
        items={[
          {
            id: '_mno97ls38',
            tweet: 'Surely crypto must be able to save us, no?'
          }
        ]}
        onChangeTweet={onChangeTweet}
        source="Surely crypto must be able to save us, no?"
      />
    );

    fireEvent.change(getByTestId('textarea'), {
      target: { value: 'Surely crypto must be able to save us, yes?' }
    });

    expect(onChangeTweet).toBeCalled();
  });
});
