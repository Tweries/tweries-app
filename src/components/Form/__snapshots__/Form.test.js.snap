// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> to match snapshot 1`] = `
<div>
  <form
    class="flex flex-col"
  >
    <div>
      ReplyToTweet
    </div>
    <p
      class="italic py-4 text-sm"
    >
      Start typing. To insert a break prior to reaching 280 characters, please use Newline(s)
    </p>
    <textarea
      class="p-2 tweries-background-color-blue-white tweries-border"
      data-testid="source"
      name="source"
      placeholder="Type your thoughts here"
      rows="8"
    />
    <p
      class="text-right text-xs text-gray-700"
    >
      Â 
    </p>
    <div>
      TweetstormButton
    </div>
  </form>
</div>
`;
